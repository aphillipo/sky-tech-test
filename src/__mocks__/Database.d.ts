export type DatabaseParams = {};

export type DatabaseResult<T> = {
  results: T[];
};

export type Database = {
  connect: (params: DatabaseParams) => Promise<boolean>;
  query: <T = any>(
    query: string,
    variables: object
  ) => Promise<DatabaseResult<T>>;
};
