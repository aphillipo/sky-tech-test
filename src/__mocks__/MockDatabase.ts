import { Database, DatabaseResult } from "./database";

export class MockDatabase implements Database {
  public constructorMock = jest.fn();

  constructor(...args: any[]) {
    this.constructorMock(...args);
  }

  connect = jest.fn((_params: any) => {
    return Promise.resolve(true);
  });

  query = jest.fn(() => {
    const newDbResult: DatabaseResult<any> = {
      results: [],
    };
    return Promise.resolve(newDbResult);
  });
}
