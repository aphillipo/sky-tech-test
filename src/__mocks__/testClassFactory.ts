import { Database } from "./Database";

export const testClassFactory = () => {
  class Test {
    // tell typescript the database will be filled in, wish this was possible without !
    fakeDatabase!: Database;

    async join() {
      return this.fakeDatabase.connect({});
    }

    async signIn(username: string, token: string) {
      // add user
      await this.fakeDatabase.query("UPDATE user SET user.name = ?", [
        username,
      ]);
      return true;
    }
  }
  return Test;
};
