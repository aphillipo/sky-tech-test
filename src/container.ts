declare type ConstructorFunction<T = any> = new (...args: any[]) => T;

type Services = {
  [serviceName: string]: unknown;
};

export class Container {
  private services: Services = {};

  private injectedSingletons: Services = {};

  static defaultContainer: Container;

  register<T = any>(serviceName: string, serviceClass: ConstructorFunction<T>) {
    this.services[serviceName] = serviceClass as T;
  }

  setDefaultContainer() {
    // set the static option, bit weird but it works
    // we could make the whole Container static but I like being able to make many and pass to inject.
    Container.defaultContainer = this;
  }

  // TODO: should probably avoid invoking this one.
  invoke<T = any>(serviceName: string, serviceOptions?: object): T {
    return new (this.services[serviceName] as ConstructorFunction<T>)(
      serviceOptions
    );
  }

  invokeOrFetchSingleton<T = any>(serviceName: string): T {
    if (!this.injectedSingletons[serviceName]) {
      this.injectedSingletons[serviceName] = this.invoke(serviceName);
    }
    return this.injectedSingletons[serviceName] as T;
  }

  getServices() {
    return this.services;
  }
}
