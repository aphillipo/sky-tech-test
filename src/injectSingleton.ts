import { Container } from "./container";

export type InjectSingletonPayload = string[];

export const injectSingleton = (
  injectables: InjectSingletonPayload,
  container?: Container
) => {
  const injectContainer = container || Container.defaultContainer;

  if (!injectContainer) {
    throw new Error(
      "Cannot find container, call setDefaultContainer to set globally or pass into inject"
    );
  }

  return (constructor: Function) => {
    injectables.forEach((serviceName: string) => {
      constructor.prototype[serviceName] =
        injectContainer.invokeOrFetchSingleton(serviceName);
    });
  };
};
