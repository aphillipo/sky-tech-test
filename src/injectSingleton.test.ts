import { expect, describe, test, beforeEach } from "@jest/globals";
import { Container } from "./container";
import { injectSingleton } from "./injectSingleton";

import { MockDatabase } from "./__mocks__/MockDatabase";
import { testClassFactory } from "./__mocks__/testClassFactory";

describe("injectSingleton", () => {
  let container: Container;

  beforeEach(() => {
    container = new Container();
    container.register("fakeDatabase", MockDatabase);
  });

  test("should return true when calling connect", async () => {
    const Test = testClassFactory();

    // decorating like this manually
    const decorateWithMockDatabase = injectSingleton(
      ["fakeDatabase"],
      container
    );
    decorateWithMockDatabase(Test);

    // create our test class
    const newTest = new Test();
    const connected = await newTest.join();

    expect(connected).toStrictEqual(true);
  });

  test("should throw when there is no container passed or global container set", () => {
    const Test = testClassFactory();

    expect(() => {
      const decorateWithMockDatabase = injectSingleton(["fakeDatabase"]);
      decorateWithMockDatabase(Test);
    }).toThrowErrorMatchingSnapshot();
  });

  test("setting container to default should avoid throw", async () => {
    const Test = testClassFactory();
    container.setDefaultContainer();

    const decorateWithMockDatabase = injectSingleton(["fakeDatabase"]);
    decorateWithMockDatabase(Test);

    const newTest = new Test();
    // this calls our injected fakeDatabase
    const success = await newTest.signIn("aphillipo", "fake-token");
    expect(success).toBe(true);
  });
});
