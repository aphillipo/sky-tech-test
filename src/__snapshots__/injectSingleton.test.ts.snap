// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`injectSingleton should throw when there is no container passed or global container set 1`] = `"Cannot find container, call setDefaultContainer to set globally or pass into inject"`;
